### DISCUSSION TITLE
How to catch OS signals and errors during setup?

### DISCUSSION THREAD

**Discussion by MatanYadaev** [vitest-dev/vitest#3745] [Q&A]
On every tests execution, I create a DB container, and I stop it once the tests execution ends.

But as of today I have two leaks that makes the container live forever:
1. On OS signals (SIGTERM, SIGKILL, basically cmd+c)
2. On error that is thrown during the setup/teardown.

**How can I make sure to run `container.stop()` when these two ^ happen?**
I tried `process.on(...)` and it didn't work.

`vite.config.ts`:
```ts
import { defineConfig } from 'vite';

export default defineConfig({
  test: {
    globalSetup: ['vitest-setup.ts'], // This creates a DB container
  },
});
```

`vitest-setup.ts`:
```ts
import Docker from 'dockerode';

let postgresContainer: Docker.Container | undefined;

export const setup = async () => {
  postgresContainer = (await createPostgresContainer())!;

  return async () => {
    await stopPostgresContainer()
  };
};

function stopPostgresContainer() {
  if (!postgresContainer) {
    return;
  }

  const containerData = await postgresContainer.inspect();

  if (!containerData.State.Running) {
    return;
  }

  await postgresContainer.stop();
}
```

Trying to catch OS signals and unhandled exceptions doesn't work:
```ts
// In `vitest-setup.ts`
['uncaughtException', 'unhandledRejection', 'SIGTERM', 'SIGKILL'].forEach((event) => {
  process.on(event, async (error) => {
    // This never gets called
    console.log('Catch error', error);
    await stopPostgresContainer()
  });
})
```

**Comment by dimisus** [comment]
Yes, thanks. I need to intercept the termination signals as well in order to clean up a fold before the next run.
